Templates/Templates/AutoDecodable.stencil

```diff
@@ -0,0 +1,37 @@
+extension KeyedDecodingContainer {
+  public func decodeOrDefault<T: Decodable>(_ type: T.Type, forKey key: Key, defaultValue: T) throws -> T {
+    guard let value: T = try decodeIfPresent(type, forKey: key) else {
+      return defaultValue
+    }
+    return value
+  }
+}
+
+{# these are a giant line because of spacing, to debug just seperate the if into lines#}
+{% macro defaultValueFor variable %}{% set typename %}{{ variable.typeName.name }}{% endset %}{% if variable|annotated:"codingDefaultValue" %}{{ variable.annotations.codingDefaultValue }}){% elif variable.isOptional %}nil){% elif variable.isArray %}[]){% elif typename == "String" %}""){% elif typename == "Int" or typename == "Double" or typename == "Float" or typename == "NSNumber" %}0){% elif typename == "Bool" %}false){% endif %}{% endmacro %}
+{% macro nameWithTransform name transform %}{% if transform == "camelToSnake" or not transform %}{{ name|camelToSnakeCase }}{% elif transform == "camelToSnake:idToUUID" %}{{ name|camelToSnakeCase|replace:"id","uuid" }}{% elif transform == "idToUUID" %}{{ name|replace:"id","uuid" }}{% endif %}{% endmacro %}
+{% macro keyedNameFor type  variable %}{% if variable|annotated:"codingKey" %}{{ variable.annotations.codingKey }}{% elif type|annotated:"codingKeyTransform" %}{% call nameWithTransform variable.name type.annotations.codingKeyTransform %}{% else %}{{ variable.name }}{% endif %}{% endmacro %}
+
+{# the keys to use to decoding #}
+{% macro codingKeys type %}
+  fileprivate enum CodableKeys: String, CodingKey {
+    {% for variable in type.allVariables where variable|!annotated:"skipCoding" %}
+    case {{ variable.name }} = "{% call keyedNameFor type variable %}"
+    {% endfor %}
+  }
+{% endmacro %}
+
+{# going to exclude protocols for now, handle it seperatly #}
+{% for type in types.all|!protocol where type.based.AutoDecodable or type|annotated:"AutoDecodable" and not type.name == "AutoCodable" and not type.name == "AutoDecodable" %}
+  // MARK: {{ type.name }}+Decodable | AutoDecodable
+  // sourcery:inline:auto:{{ type.name }}.AutoDecodable
```

Comment:
I'm currently working on such template myself and I noticed that when generating constructor in the type body Swift stops to generate member wise constructors, which may just break code that was relying on such constructor to be generated. So I would suggest to not use inline here and instead generate this constructor in extension of type.

Reply:
When I was trying to do that the compiler was complaining about required constructors outside the class body (and I believe the decoding constructor has to be required?), but I'll see if theres a way around that.

Reply:
I think this is only needed for classes, for value types `required` is not needed. So I guess classes should be handled separately, with inheritance in mind for which there is superDecoder/Encoder, as far as I know